examples/experimental/native/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,199 @@
+
+import data
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+class Generator(nn.Module):
+    def __init__(self, latent_dim, img_shape):
+        super(Generator, self).__init__()
+        self.img_shape = img_shape
+
+        def block(in_feat, out_feat, normalize=True):
+            layers = [nn.Linear(in_feat, out_feat)]
+            if normalize:
+                layers.append(nn.BatchNorm1d(out_feat, 0.8))
+            layers.append(nn.LeakyReLU(0.2, inplace=True))
+            return layers
+
+        self.model = nn.Sequential(
+            *block(latent_dim, 128, normalize=False),
+            *block(128, 256),
+            *block(256, 512),
+            *block(512, 1024),
+            nn.Linear(1024, int(np.prod(img_shape))),
+            nn.Tanh()
+        )
+
+    def forward(self, z):
+        img = self.model(z)
+        img = img.view(img.size(0), *self.img_shape)
+        return img
+
+
+class Discriminator(nn.Module):
+    def __init__(self, img_shape):
+        super(Discriminator, self).__init__()
+
+        self.model = nn.Sequential(
+            nn.Linear(int(np.prod(img_shape)), 512),
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(512, 256),
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(256, 1),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, img):
+        img_flat = img.view(img.size(0), -1)
+        validity = self.model(img_flat)
+
+        return validity
+
+
+class GAN(nn.Module):
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
+        super(GAN, self).__init__()
+
+        self.context = trial_context
+
+        mnist_shape = (1, 28, 28)
+        self.generator = Generator(latent_dim=self.context.get_hparam("latent_dim"), img_shape=mnist_shape)
+        self.discriminator = Discriminator(img_shape=mnist_shape)
+
+    def forward(self, z):
+        return self.generator(z)
+
+
+class MNistTrial(PyTorchTrial):
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
+        self.context = trial_context
+        self.logger = TorchWriter()
+
+        # Create a unique download directory for each rank so they don't overwrite each other.
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_downloaded = False
+
+        self.model = self.context._Model(GAN(trial_context))
+
+        lr = self.context.get_hparam("lr")
+        b1 = self.context.get_hparam("b1")
+        b2 = self.context.get_hparam("b2")
+
+        self.opt_g = self.context._Optimizer(torch.optim.Adam(self.model.generator.parameters(), lr=lr, betas=(b1, b2)))
+        self.opt_d = self.context._Optimizer(torch.optim.Adam(self.model.discriminator.parameters(), lr=lr, betas=(b1, b2)))
+
+        # (self.model.generator, self.model.discriminator), (self.opt_g, self.opt_d) = self.context._configure_apex_amp(
```

Comment:
Unfortunately, `F. binary_cross_entropy ` doesn't work with apex.amp directly. I might need to remove this line.